In [3]:
import pickle
%matplotlib inline
from collections import defaultdict
import pylab as plt
from os import listdir

def get_clusters(part):
    cluster_assignment=defaultdict(list)
    for k,v in part.iteritems():
        cluster_assignment[v].append(k)
    #print out the length of each cluster
    for k in cluster_assignment.keys():
        print k,len(cluster_assignment[k])
    return cluster_assignment



def read_csv_data_meta_unigram(data_file):
    f=open(data_file,'r').read().split('\n')
    all_csv_data=[]
    all_lab=[]
    for line in f:
        if line=='':
            continue
        l=line.split(',')[:-1]
        b=line.split(',')[-1]
        #l=[float(i) for i in l]
        all_csv_data.append(l)
        all_lab.append(b)
    return all_csv_data,all_lab


def read_csv_data_meta(data_file):
    f=open(data_file,'r').read().split('\n')
    all_csv_data=[]
    all_lab=[]
    for line in f:
        if line=='':
            continue
        l=line.split(',')[:-4]
        b=line.split(',')[-4:]
        #l=[float(i) for i in l]
        all_csv_data.append(l)
        all_lab.append(b)
    return all_csv_data,all_lab


def plot_clusters(path,pfile,csv_file):
    #pfile='downsample_syl_3_meta_200_421_pkl_part.pkl'
    #path='core_subdata/'
    part=pickle.load(open(path+pfile,'r'))
    clust=get_clusters(part)
    #csv_file='subdata/downsample_syl_3_meta_200_421.csv'
    ngram=csv_file.split('.')[0].split('_')[-1]
    csvdata,lab=read_csv_data_meta(csv_file)
    #print len(csvdata)
    for j in clust.keys():
        inds=clust[j]
        plt.figure()
        for i in inds:
            #print 'i',i
            
            #print len(csvdata[i])
            plt.plot(csvdata[i])
            plt.title(ngram+"|"+str(j))

In [1]:
def get_classes(path,pfile,csv_file):
    part=pickle.load(open(path+pfile,'r'))
    clust=get_clusters(part)
    ngram=csv_file.split('.')[0].split('_')[-1]
    csvdata,lab=read_csv_data_meta(csv_file)
    all_data=[]
    for key in clust.keys():
        inds=clust[key]
        if len(inds)<50:
            print 'key excluded:',key
            continue
        for i in inds:
            data=csvdata[i]
            labels=lab[i]
            data.extend(labels)
            data.append(str(key))
            all_data.append(data)
    return all_data
            
            
def write_file(data,pfile):
    outfile=pfile.replace('.pkl','_class.csv')
    f=open(outfile,'w').close()
    f=open(outfile,'a')
    num_col=len(data[0])
    header=[str(i+1) for i in range(num_col-1)]
    header.append('label')
    f.write(','.join(header)+'\n')
    for line in data:
        f.write(','.join(line)+'\n')
    f.close()         
    
    

# write to a csv file where each line has a label indicating the class the the shape of the ts belongs to

In [7]:
path='community_partition/core_subdata/'
pfiles = [ f for f in listdir(path) if f.endswith('_part.pkl')]


In [9]:
len(pfiles)

86

In [26]:
#demo
pfile=pfiles[0]
csv_file='subdata/'+pfile.replace('_pkl_part.pkl','.csv')
d=get_classes(path,pfile,csv_file)

0 181
1 183
2 212
3 116
4 2
5 4
6 9
7 5
8 2
9 12
10 12
11 2
12 2
13 11
14 4
15 2
16 2
17 6
18 2
19 3
20 2
21 2
22 2
23 2
24 5
25 2
key excluded: 4
key excluded: 5
key excluded: 6
key excluded: 7
key excluded: 8
key excluded: 11
key excluded: 12
key excluded: 14
key excluded: 15
key excluded: 16
key excluded: 17
key excluded: 18
key excluded: 19
key excluded: 20
key excluded: 21
key excluded: 22
key excluded: 23
key excluded: 24
key excluded: 25


In [44]:
#process all
for pfile in pfiles:
    csv_file='subdata/'+pfile.replace('_pkl_part.pkl','.csv')
    d=get_classes(path,pfile,csv_file)
    write_file(d,pfile)

0 181
1 183
2 212
3 116
4 2
5 4
6 9
7 5
8 2
9 12
10 12
11 2
12 2
13 11
14 4
15 2
16 2
17 6
18 2
19 3
20 2
21 2
22 2
23 2
24 5
25 2
key excluded: 4
key excluded: 5
key excluded: 6
key excluded: 7
key excluded: 8
key excluded: 9
key excluded: 10
key excluded: 11
key excluded: 12
key excluded: 13
key excluded: 14
key excluded: 15
key excluded: 16
key excluded: 17
key excluded: 18
key excluded: 19
key excluded: 20
key excluded: 21
key excluded: 22
key excluded: 23
key excluded: 24
key excluded: 25
0 85
1 119
2 154
3 17
4 168
5 175
6 11
7 2
8 11
9 2
10 3
11 2
12 4
key excluded: 3
key excluded: 6
key excluded: 7
key excluded: 8
key excluded: 9
key excluded: 10
key excluded: 11
key excluded: 12
0 57
1 53
2 120
3 112
4 119
5 80
6 2
7 3
8 2
9 11
10 3
11 3
12 2
key excluded: 6
key excluded: 7
key excluded: 8
key excluded: 9
key excluded: 10
key excluded: 11
key excluded: 12
0 82
1 136
2 38
3 173
4 2
5 58
6 5
7 172
8 9
9 12
10 2
11 6
12 3
13 3
14 7
15 2
16 2
17 2
18 2
19 4
20 2
21 2
22 3
23 3
24 

# 

# data written to file

note that the data will be smaller than the original csv file - some TS are excluded because some of the cluster keys are excluded, because they are too small a cluster - we consider them as outliers.

# classification using ts data

we used ts ngram data to classify into these categories (classes). It is able to do it with RF in general with high accuracy between 85% to 90%. 

next we can proceed to inter-domain classification. you can include starting and ending pitch as a feature. but we want in general the out of domain features. for data, we need to map between the text and these TS.